src/vring.rs

```diff
@@ -0,0 +1,239 @@
+// Copyright 2019 Intel Corporation. All Rights Reserved.
+// Copyright 2021 Alibaba Cloud Computing. All rights reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+//! Struct to maintain state information and manipulate vhost-user queues.
+
+use std::fs::File;
+use std::io;
+use std::os::unix::io::{FromRawFd, IntoRawFd};
+use std::sync::{Arc, RwLock, RwLockReadGuard, RwLockWriteGuard};
+
+use virtio_queue::Queue;
+use vm_memory::{GuestAddress, GuestAddressSpace, GuestMemoryAtomic, GuestMemoryMmap};
+use vmm_sys_util::eventfd::EventFd;
+
+/// Struct to maintain raw state information for a vhost-user queue.
+pub struct VringState<M: GuestAddressSpace> {
+    queue: Queue<M>,
+    kick: Option<EventFd>,
+    call: Option<EventFd>,
+    err: Option<EventFd>,
+    enabled: bool,
+}
+
+impl<M: GuestAddressSpace> VringState<M> {
+    fn new(mem: M, max_queue_size: u16) -> Self {
+        VringState {
+            queue: Queue::new(mem, max_queue_size),
+            kick: None,
+            call: None,
+            err: None,
+            enabled: false,
+        }
+    }
+
+    /// Get a mutable reference to the underlying raw `Queue` object.
+    pub fn get_queue_mut(&mut self) -> &mut Queue<M> {
+        &mut self.queue
+    }
+
+    /// Get a immutable reference to the kick event fd.
+    pub fn get_kick(&self) -> &Option<EventFd> {
+        &self.kick
+    }
+}
+
+/// Struct to maintain state information and manipulate a vhost-user queue.
+#[derive(Clone)]
+pub struct Vring<M: GuestAddressSpace = GuestMemoryAtomic<GuestMemoryMmap>> {
+    state: Arc<RwLock<VringState<M>>>,
+}
+
+impl<M: GuestAddressSpace> Vring<M> {
```

Comment:
Do you think we should add wrappers for `enable_notification`, `disable_notification` and `needs_notification` too?

Reply:
sure, will take care of them.

Reply:
fixed